src/abstractdatatable/join.jl

```diff
@@ -214,6 +144,199 @@ function sharepools(dt1::AbstractDataTable, dt2::AbstractDataTable)
+        on_cols = (isa(on, Symbol) ? fill(on::Symbol, 1) : on)::Vector{Symbol}
+        new(dtl, dtr, dtl[on_cols], dtr[on_cols], on_cols)
+    end
+end
+
+_DataTableJoiner{DT1<:AbstractDataTable, DT2<:AbstractDataTable}(dtl::DT1, dtr::DT2, on::Union{Symbol,Vector{Symbol}}) =
+    _DataTableJoiner{DT1,DT2}(dtl, dtr, on)
+
+# helper map between the row indices in original and joined frame
+immutable _RowIndexMap
+  orig::Vector{Int} # row indices in the original frame
+  join::Vector{Int} # row indices in the resulting joined frame
+end
+
+Base.length(x::_RowIndexMap) = length(x.orig)
+
+# fix the result of the rightjoin by taking the nonnull values from the right frame
+function fix_rightjoin_column!(res_col::AbstractArray, col_ix::Int, joiner::_DataTableJoiner,
+                              all_orig_left_ixs::Vector{Int}, rightonly_ixs::_RowIndexMap)
+    res_col[rightonly_ixs.join] = joiner.dtr_on[rightonly_ixs.orig, col_ix]
+    res_col
+end
+
+# since setindex!() for PoolDataArray is very slow,
+# it requires special handling
+function fix_rightjoin_column!{T,N}(res_col::Union{AbstractCategoricalArray{T,N}, AbstractNullableCategoricalArray{T,N}},
+                                    col_ix::Int, joiner::_DataTableJoiner,
+                                    all_orig_left_ixs::Vector{Int}, rightonly_ixs::_RowIndexMap)
+    left_col = joiner.dtl_on[col_ix]
+    right_col = joiner.dtr_on[col_ix]
+    if levels(left_col) == levels(right_col)
+        res_col.refs[rightonly_ixs.join] = right_col.refs[rightonly_ixs.orig]
+        return res_col
+    end
+    # merge the pools
+    # FIXME use sharedpools()?
+    newlevels, ordered = CategoricalArrays.mergelevels(levels(left_col), levels(right_col))
+    if length(newlevels) <= typemax(reftype(left_col))
+        newreftype = reftype(left_col)
+    elseif length(newlevels) <= typemax(reftype(right_col))
+        newreftype = reftype(right_col)
+    else
+        newreftype = reftype(length(newlevels))
+    end
+    new_refs = newreftype[
+            indexin(CategoricalArrays.index(left_col.pool), newlevels)[left_col.refs[all_orig_left_ixs]];
+            indexin(CategoricalArrays.index(right_col.pool), newlevels)[right_col.refs[rightonly_ixs.orig]]]
+    NullableCategoricalArray{T,N,newreftype}(new_refs, CategoricalPool(newlevels, ordered))
+end
+
+# composes the joined data frame using the maps between the left and right
+# frame rows and the indices of rows in the result
+function compose_joined_frame(joiner::_DataTableJoiner,
+                left_ixs::_RowIndexMap, leftonly_ixs::_RowIndexMap,
+                right_ixs::_RowIndexMap, rightonly_ixs::_RowIndexMap)
+    @assert length(left_ixs) == length(right_ixs)
+    # compose left half of the result taking all left columns
+    # FIXME is it still relevant? complicated way to do vcat that avoids expensive setindex!() for PooledDataVector
+    all_orig_left_ixs = [left_ixs.orig; leftonly_ixs.orig]
```

Comment:
AFAIR `setindex(PooledDataVector)` was very slow, because it did linear search in the pool for the value being assigned. It should not be the case for CategoricalArray. The problem is that now I don't remember anymore how the simple version should look like. :) Need to get into the code once again.